<div align="right" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/ExploreAI_logos/Logo blue_dark.png"  style="width:25px" align="right";/>
</div>

## Instructions to students

This challenge is designed to determine how much you have learned so far and will test your knowledge on SQL.

The answers for this challenge should be selected on Athena for each corresponding multiple-choice question. The questions are included in this notebook and are numbered according to the Athena questions. The options for each question have also been included.

Do not add or remove cells in this notebook. Do not edit or remove the `%%sql` comment as it is required to run each cell.

**_Good luck!_**

## Honour code

I,Mgcini Emmanuel Majola, confirm – by submitting this document – that the solutions in this notebook are a result of my own work and that I abide by the EDSA honour code (https://drive.google.com/file/d/1bl28j1Qe5jNyfnd7Tcuyilj-NBEQP_Tt/view?usp=sharing).

Non-compliance with the honour code constitutes a material breach of contract. 

>⚠️ Sharing or uploading this assessment as whole, or in part is a violation of the honour code.

## The dataset

The dataset comprises several tables related to water management and infrastructure, primarily focusing on various aspects such as water sources, treatment plants, and reservoirs. Each table contains essential information regarding different components involved in managing the water resources and sanitation.

Below is an Entity Relationship Diagram (ERD) of the `Water_sanitation` database:

#  

<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/water_sanitation_ERD.png"  style="width:650px";/>
</div>

This ER diagram illustrates the relationships between the tables in the dataset.

## Loading the database

Before we begin, we need to prepare our SQL environment.

We can do this by loading the magic command `%load_ext sql`.

In [1]:
# Load and activate the SQL extension to allow us to execute SQL in a Jupyter notebook. 
# If you get an error here, make sure that mysql and pymysql are installed correctly. 

%load_ext sql

Now we can load our database. 

To do this, we will need to ensure we have downloaded the `Water_sanitation.db` sqlite file and have stored it in a known location.

[Download Water_sanitation.db](https://github.com/Explore-AI/Public-Data/raw/master/Water_sanitation.db)

In [2]:
%sql sqlite:///Water_sanitation.db

We can also load our database using a sqlite3 connection for those commands that are not supported by Ipython.

In [3]:
import sqlite3
# Connect to the Water_sanitation database
conn = sqlite3.connect('Water_sanitation.db')

In [9]:
cursor = conn.cursor()

# Questions

Use the given cell below each question to execute the SQL queries to find the correct input from the options provided. Your solution should match one of the multiple-choice questions on Athena.

### Question 1
As data analysts, we are preparing a report on urbanisation trends, and we need to calculate the average population for all the cities.

What is the average population size (rounded to 2 decimal places) of all the cities in our dataset?

#### Options

* 2578203.17
* 5371256.60
* 3455456.32
* 4889967.52

In [4]:
#Your code here
cursor.execute("SELECT AVG(Population) FROM City")
cursor.fetchall()


NameError: name 'cursor' is not defined

### Question 2

A city planner is prioritising infrastructure projects and needs to identify the top 5 cities with the highest population.

Which SQL query correctly retrieves the names of the top 5 cities with the highest population, ordered by population in descending order?

#### Options:

* `SELECT TOP 5 CityName, Population`<br> 
`FROM City`<br>
`ORDER BY Population DESC;`<br>

* `SELECT CityName, Population`<br> 
`FROM City`<br>
`ORDER BY Population DESC LIMIT 5;`<br>

* `SELECT CityName, Population`<br>
`FROM City`<br> 
`ORDER BY Population DESC FETCH FIRST 5 ROWS ONLY;`<br>

* `SELECT CityName, Population `<br>
`FROM City`<br> 
`ORDER BY Population LIMIT 5;`<br>


In [22]:
#Your code here
cursor.execute("SELECT CityName, Population FROM City ORDER BY Population DESC LIMIT 5")
cursor.fetchall()

[('Port harcourt', 9767493),
 ('Gaborone', 9434152),
 ('Soweto', 9307546),
 ('Suez', 9285971),
 ('Abuja', 9152131)]

### Question 3

A city planner is analysing water usage patterns and needs to find the total volume of water used by each city, ordered from the highest usage to the lowest.

Which city has the highest water usage?

#### Options
* Johannesburg
* Abidjan
* Lagos
* Kampala

In [34]:
#Your code here
cursor.execute("SELECT DISTINCT cityname,SUM(Volume) FROM city JOIN watersource ON city.cityid = watersource.cityid JOIN waterusage ON watersource.sourceid = waterusage.sourceid WHERE cityname IN ('Johannesburg','Abidjan','Lagos','Kampala') GROUP BY cityname ORDER BY volume DESC")
cursor.fetchall()

[('Lagos', 1328885.2900000003),
 ('Abidjan', 1621934.4),
 ('Kampala', 1410650.9200000004),
 ('Johannesburg', 805500.15)]

### Question 4
The government wants to identify the top five cities in **Southern Africa** with a population **greater than 1 million**. 

Which cities fall within this category?

#### Options

* Giza, Johannesburg, Abidjan, Cape Town, Addis Ababa
* Brazzaville, Mombasa, Addis Ababa, Bloemfontein, Giza
* Kampala, Cape Town, Abidjan, Freetown, Johannesburg
* Kampala, Johannesburg, Ababa, Cape, Giza



In [42]:
#Your code here
cursor.execute("SELECT CityName, Population, regionname FROM City JOIN region ON city.regionid = region.regionid WHERE regionname LIKE '%South%' AND Population > 1000000 ")
cursor.fetchall()

[('Giza', 5115607, 'South'),
 ('Johannesburg', 5089219, 'South'),
 ('Abidjan', 4840111, 'Southeast'),
 ('Cape town', 3475596, 'Southeast'),
 ('Addis ababa', 6214148, 'Southwest'),
 ('Mombasa', 6505433, 'South'),
 ('Kampala', 2860610, 'Southeast'),
 ('Bloemfontein', 5820903, 'Southeast'),
 ('Brazzaville', 8491153, 'Southwest'),
 ('Freetown', 5070331, 'South')]

### Question 5

A social services agency tasked us with categorising cities based on population size for resource allocation purposes.

They'd like to categorise the cities according to the following population thresholds: <br>
* Populations less than 1,000,000 are considered `Small`.<br>
* Populations greater than 1,000,000 and less than 5,000,000 are considered `Medium`.<br>
* Otherwise, the population is classified as `Large`.

What is the population category for **Kinshasa**?

### Options

* Small<br>
* Medium<br>
* Large<br>
* None<br>

In [11]:
#Your code here
cursor.execute("SELECT CityName, CASE WHEN Population < 1000000 THEN 'Small' WHEN Population BETWEEN 1000000 AND 5000000 THEN 'medium' ELSE 'Large' END AS [Population Threshold] FROM City WHERE Cityname = 'Kinshasa'")
cursor.fetchall()

[('Kinshasa', 'Large')]

### Question 6
The social services agency decides to further categorise population sizes by including  water usage patterns. They want to categorise cities based on their water usage and population. The new categories will look as follows:

* Category A: Cities with a population greater than 1 million and an average water usage volume per month greater than 500,000 gallons are classified as Category A cities. These cities have both a large population and high water usage.
* Category B: Cities with a population between 500,000 and 1 million, inclusive, or an average water usage volume per month between 300,000 and 500,000 gallons, inclusive, are classified as Category B cities. These cities have either a moderate population or moderate water usage, but not both.
* Category C: All other cities that do not meet the criteria for Category A or B are classified as Category C cities. These cities have a relatively lower population and water usage compared to Category A and B cities.

What is the population category for **Kinshasa**?

### Options
* Category A
* Category B
* Category C
* None of the above

In [45]:
#Your code here
cursor.execute("SELECT DISTINCT cityname, CASE WHEN Population > 1000000 AND Volume > 500000 THEN 'Category A' WHEN (Population BETWEEN 500000 AND 1000000) AND (Volume BETWEEN 300000 AND 500000) THEN 'Category B' ELSE 'Category C' END AS [Population Threshold]  FROM city JOIN watersource ON city.cityid = watersource.cityid JOIN waterusage ON watersource.sourceid = waterusage.sourceid WHERE cityname = 'Kinshasa'")
cursor.fetchall()

[('Kinshasa', 'Category C')]

### Question 7
A city planner needs to identify cities with populations below the average population of all cities.

Write a SQL query that correctly retrieves the names of cities with populations below the average population of all cities.

True or false? Abuja falls within this list of cities.

**Hint: It will be easier to find the cities if they're ordered.**

#### Options
* True  <br>
* False

In [51]:
#Your code here
cursor.execute("SELECT Cityname, Population FROM City WHERE population < 5371256.60 ORDER BY Cityname")
cursor.fetchall()

[('Abidjan', 4840111),
 ('Antananarivo', 2586262),
 ('Beira', 3550905),
 ('Cape town', 3475596),
 ('Casablanca', 3998673),
 ('Conakry', 2594819),
 ('Dar es salaam', 3669154),
 ('Durban', 4842918),
 ('Freetown', 5070331),
 ('Giza', 5115607),
 ('Harare', 3244159),
 ('Johannesburg', 5089219),
 ('Kampala', 2860610),
 ('Khartoum', 1026664),
 ('Kumasi', 1148956),
 ('Lagos', 138638),
 ('Lilongwe', 1992252),
 ('Luanda', 689256),
 ('Lusaka', 545084),
 ('Maiduguri', 2087623),
 ('Maseru', 2419921),
 ('Mbuji-mayi', 4202841),
 ("N'djamena", 4849424),
 ('Yaoundé', 3577719)]

### Question 8

A researcher is analysing water quality data and needs to calculate the average chlorine level for each city.

Which SQL query correctly calculates the average chlorine level for each city using a Common Table Expression (CTE)?

#### Options
* `WITH AvgChlorine AS (`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`SELECT Source.CityID, City.CityName, AVG(WaterQuality.ChlorineLevel) AS AvgChlorine`<br> 
    &nbsp;&nbsp;&nbsp;&nbsp;`FROM WaterQuality` <br>
    &nbsp;&nbsp;&nbsp;&nbsp;`JOIN WaterSource Source ON WaterQuality.SourceID = Source.SourceID`<br> 
    &nbsp;&nbsp;&nbsp;&nbsp;`JOIN City ON Source.CityID = City.CityID` <br>
&nbsp;&nbsp;&nbsp;&nbsp;`)`<br><br>
`SELECT CityID, CityName, AvgChlorine FROM AvgChlorine;`<br>

* `WITH AvgChlorine AS (`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`SELECT Source.CityID, City.CityName, AVG(WaterQuality.ChlorineLevel) AS AvgChlorine` 
    &nbsp;&nbsp;&nbsp;&nbsp;`FROM WaterQuality ` <br>
    &nbsp;&nbsp;&nbsp;&nbsp;`JOIN City ON Source.CityID = City.CityID` <br>
    &nbsp;&nbsp;&nbsp;&nbsp;`GROUP BY Source.CityID, City.CityName`<br>
&nbsp;&nbsp;&nbsp;&nbsp;`)`<br><br>
`SELECT CityID, CityName, AvgChlorine FROM AvgChlorine;`

* `WITH AvgChlorine AS (`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`SELECT Source.CityID, City.CityName, AVG(WaterQuality.ChlorineLevel) AS AvgChlorine` <br>
    &nbsp;&nbsp;&nbsp;&nbsp;`FROM WaterQuality`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`JOIN WaterSource Source ON WaterQuality.SourceID = Source.SourceID` <br>
    &nbsp;&nbsp;&nbsp;&nbsp;`JOIN City ON Source.CityID = City.CityID`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`GROUP BY Source.CityID, City.CityName`<br>
&nbsp;&nbsp;&nbsp;&nbsp;`)`<br><br>
`SELECT CityID, CityName, AvgChlorine FROM AvgChlorine;`<br>

* `WITH AvgChlorine AS (`<br>
&nbsp;&nbsp;&nbsp;&nbsp;`SELECT CityID, AVG(ChlorineLevel) AS AvgChlorine` <br>
&nbsp;&nbsp;&nbsp;&nbsp;`FROM WaterQuality GROUP BY CityID`<br>
&nbsp;&nbsp;&nbsp;&nbsp;`)`<br><br>
`SELECT CityID, CityName, AvgChlorine FROM AvgChlorine;`



In [12]:
cursor.execute("""WITH AvgChlorine AS (
    SELECT Source.CityID, City.CityName, AVG(WaterQuality.ChlorineLevel) AS AvgChlorine
    FROM WaterQuality
    JOIN WaterSource Source ON WaterQuality.SourceID = Source.SourceID
    JOIN City ON Source.CityID = City.CityID
    GROUP BY Source.CityID, City.CityName
    )

SELECT CityID, CityName, AvgChlorine FROM AvgChlorine""")
cursor.fetchall()

[(1, 'Cairo', 0.8195833333333336),
 (2, 'Lagos', 0.7720833333333333),
 (3, 'Kinshasa', 0.8029166666666668),
 (4, 'Giza', 0.8733333333333332),
 (5, 'Nairobi', 0.7770833333333332),
 (6, 'Johannesburg', 0.8374999999999998),
 (7, 'Abidjan', 0.87),
 (8, 'Casablanca', 0.8283333333333335),
 (9, 'Cape town', 0.8329166666666664),
 (10, 'Durban', 0.7420833333333334),
 (11, 'Alexandria', 0.9145833333333334),
 (12, 'Addis ababa', 0.8529166666666667),
 (13, 'Dar es salaam', 0.8158333333333333),
 (14, 'Accra', 0.9054166666666666),
 (15, 'Khartoum', 0.9070833333333335),
 (16, 'Suez', 0.9370833333333336),
 (17, 'Mombasa', 0.88375),
 (18, 'Kampala', 0.8224999999999998),
 (19, 'Maputo', 0.8566666666666666),
 (20, 'Port harcourt', 0.9091666666666663),
 (21, 'Ouagadougou', 0.8145833333333332),
 (22, 'Lusaka', 0.9029166666666666),
 (23, 'Abuja', 0.7937500000000001),
 (24, 'Bamako', 0.9229166666666667),
 (25, 'Kano', 0.8866666666666668),
 (26, 'Dakar', 0.7608333333333336),
 (27, 'Cotonou', 0.809166666666666

### Question 9
A water management agency is conducting a study on the usage of different water sources across cities. They want to calculate the total volume of water used from each water source type and present the results in descending order of total volume.

What is the order of the source types?

#### Options
* Wastewater, Stormwater, Bore water, Groundwater
* Bore water, Wastewater, Stormwater, Groundwater
* Groundwater, Stormwater, Wastewater, Bore water
* Stormwater, Wastewater, Bore water, Groundwater


In [61]:
cursor.execute("SELECT Sourcetype, SUM(Volume) AS [Total Volume] FROM watersource JOIN Waterusage ON watersource.sourceid = waterusage.sourceid GROUP BY sourcetype ORDER BY [Total Volume] DESC ")
cursor.fetchall()

[('Wastewater', 16739327.739999987),
 ('Stormwater', 16282624.900000002),
 ('Bore water', 13652285.320000008),
 ('Groundwater', 10907783.52)]

### Question 10
A water management agency is analysing the water quality data across different cities to identify cities with the highest average pH levels in their water sources. They need to list the top 5 cities with the highest average pH levels in descending order.

What are the top 5 cities with the highest average pH levels?

#### Options
* Cairo, Lagos, Kinshasa, Giza, Nairobi
* Alexandria, abidjan,Suez, Harare, Freetown
* Kinshasa, Conakry, Dar es Salaam, N'Djamena, Antananarivo
* Conakry, Mbuji-Mayi, Kinshasa, Dar es Salaam, Luanda


In [64]:
cursor.execute("""SELECT DISTINCT cityname,AVG(ph) AS PH
FROM city JOIN watersource ON city.cityid = watersource.cityid 
JOIN waterquality ON watersource.sourceid = waterquality.sourceid 
GROUP BY cityname ORDER BY ph DESC""")
cursor.fetchall()

[('Kinshasa', 7.568750000000001),
 ('Conakry', 7.559999999999999),
 ('Dar es salaam', 7.514583333333332),
 ("N'djamena", 7.511666666666667),
 ('Antananarivo', 7.497083333333333),
 ('Bamako', 7.406666666666665),
 ('Ouagadougou', 7.4054166666666665),
 ('Ibadan', 7.4004166666666675),
 ('Mbuji-mayi', 7.387916666666666),
 ('Monrovia', 7.376666666666666),
 ('Lilongwe', 7.350833333333334),
 ('Maseru', 7.350833333333333),
 ('Yaoundé', 7.335833333333334),
 ('Kisangani', 7.323750000000001),
 ('Johannesburg', 7.322083333333334),
 ('Kano', 7.319583333333331),
 ('Port harcourt', 7.316666666666669),
 ('Luanda', 7.316250000000004),
 ('Maputo', 7.314166666666669),
 ('Gaborone', 7.305000000000001),
 ('Abuja', 7.297916666666667),
 ('Soweto', 7.2825),
 ('Cairo', 7.258333333333333),
 ('Mombasa', 7.250833333333333),
 ('Lubumbashi', 7.246666666666666),
 ('Accra', 7.242916666666667),
 ('Lagos', 7.240833333333335),
 ('Casablanca', 7.2391666666666685),
 ('Maiduguri', 7.223333333333332),
 ('Nairobi', 7.21916666

### Question 11

A water treatment company wants to analyse the capacity utilisation of water treatment plants. They need to calculate the utilisation rate for each water treatment plant and list them in descending order of utilisation rate.

`UtilisationRate = Total Volume of Water Used / Capacity of Water Treatment Plant * 100
`
<br>
* **Note that the Capacity column in the WaterTreatmentPlant table is in millions.**

What are the top 3 plants with the highest utilisation rate?

### Options

* Abidja_BEB1AE, Abuj_D3F7DB, Accr_A094F3
* Johannesbur_545E76, Brazzavill_9F1E1E, Freetow_433D56
* Durba_3BBAB6, Cotono_04562C, Abidja_BEB1AE
* Lago_F9A053, Lusak_6368E8, Maser_2A3D7E


In [72]:
cursor.execute("""SELECT DISTINCT plantname,SUM(Volume), SUM(Capacity),SUM(Volume) / SUM(Capacity) * 100 AS UtilisationRate 
FROM city JOIN watersource ON city.cityid = watersource.cityid 
JOIN waterusage ON watersource.sourceid = waterusage.sourceid 
JOIN WaterTreatmentPlant
ON city.cityid = watertreatmentplant.cityid
GROUP BY Plantname
ORDER BY UtilisationRate DESC
""")
cursor.fetchall()

[('Durba_3BBAB6', 1203058.12, 24.84745193021761, 4841776.62715158),
 ('Cotono_04562C', 1181983.54, 33.58472880094046, 3519407.7254746254),
 ('Abidja_BEB1AE', 1621934.4, 49.02611051726659, 3308307.3139745975),
 ('Lago_F9A053', 1328885.2900000003, 41.77390574003271, 3181137.282852881),
 ('Mombas_D3719A', 1293282.36, 42.48943096731768, 3043774.252930749),
 ('Lusak_6368E8', 1011616.5200000003, 35.327394887765365, 2863546.8967182315),
 ('Maser_2A3D7E', 1278851.88, 46.507137764002664, 2749797.002106317),
 ('Sowet_DA56B2', 1186427.5699999998, 47.4815463980401, 2498712.9948425014),
 ('Conakr_6D62A1', 1375500.1400000004, 59.39354730816091, 2315908.381197163),
 ('Maidugur_4D483C',
  1248583.4099999997,
  58.917592189106536,
  2119203.048883003),
 ('Accr_A094F3', 1204667.9999999998, 58.14219231598605, 2071934.2563709612),
 ("N'Djamen_14ADBF",
  1136478.9100000001,
  58.322363167356166,
  1948616.0167050692),
 ('Nairob_9E564C', 1026206.2999999999, 56.12942935660196, 1828285.645806761),
 ('Addis Ab

In [19]:
cursor.execute("""SELECT
    wtp.plantname,
    (SUM(wu.volume) / (wtp.capacity * 1000000)) * 100 AS UtilisationRate
FROM
    WaterTreatmentPlant wtp
JOIN
    WaterSource ws ON wtp.cityid = ws.cityid
JOIN
    WaterUsage wu ON ws.sourceid = wu.sourceid
GROUP BY
    wtp.plantid, wtp.plantname, wtp.capacity
ORDER BY
    UtilisationRate DESC
LIMIT 3
""")
cursor.fetchall()

[('Durba_3BBAB6', 116.20263905163783),
 ('Cotono_04562C', 84.46578541139098),
 ('Abidja_BEB1AE', 79.39937553539032)]

### Question 12

A water treatment company wants to analyse the distribution of water treatment plant capacities across different cities. They need to calculate the average capacity of water treatment plants in each city and round the results to the nearest whole number.

What is the average capacity of the water treatment plant in Maputo?

### Options

* 5
* 7
* 2
* 4


In [73]:
cursor.execute("""SELECT DISTINCT cityname, AVG(Capacity)
FROM city
JOIN WaterTreatmentPlant
ON city.cityid = watertreatmentplant.cityid
WHERE cityname = 'Maputo'

""")
cursor.fetchall()

[('Maputo', 7.135914275515422)]

In [20]:
cursor.execute("""SELECT
    c.cityname,
    ROUND(AVG(wtp.capacity)) AS AverageCapacity
FROM
    City c
JOIN
    WaterTreatmentPlant wtp ON c.cityid = wtp.cityid
GROUP BY
    c.cityname
HAVING
    c.cityname = 'Maputo';

""")
cursor.fetchall()

[('Maputo', 7.0)]

### Question 13

A water quality control agency wants to identify cities with consistently high water quality over the past year. They need to calculate the average pH level for each city and then rank the cities based on their average pH levels.

Which SQL query correctly calculates the average pH level for each city and ranks the cities based on their average pH levels?

### Options
* `SELECT CityName, AVG(pH) AS AvgpH,`<br>
       &nbsp;&nbsp;&nbsp;&nbsp;`RANK() OVER (ORDER BY AVG(pH)) AS pH_Rank`<br>
`FROM WaterQuality`<br>
`JOIN WaterSource ON WaterQuality.SourceID = WaterSource.SourceID`<br>
`JOIN City ON WaterSource.CityID = City.CityID`<br>

* `SELECT CityName, AVG(pH) AS AvgpH,`<br>
       &nbsp;&nbsp;&nbsp;&nbsp;`RANK() OVER (ORDER BY AVG(pH) DESC) AS pH_Rank`<br>
`FROM WaterQuality`<br>
`JOIN WaterSource ON WaterQuality.SourceID = WaterSource.SourceID`<br>
`JOIN City ON WaterSource.CityID = City.CityID`<br>
`GROUP BY CityName;`<br>

* `SELECT CityName, AVG(pH) AS AvgpH,`<br>
&nbsp;&nbsp;&nbsp;&nbsp;`ROW_NUMBER() OVER (PARTITION BY CityID ORDER BY AVG(pH) DESC) AS pH_Rank`<br>
`FROM WaterQuality`<br>
`JOIN WaterSource ON WaterQuality.SourceID = WaterSource.SourceID`<br>
`JOIN City ON WaterSource.CityID = City.CityID`<br>
`GROUP BY CityName;`<br>

* `SELECT CityName, AVG(pH) AS AvgpH,`<br>
       &nbsp;&nbsp;&nbsp;&nbsp;`RANK() OVER (PARTITION BY CityName ORDER BY AVG(pH) DESC) AS pH_Rank`<br>
`FROM WaterQuality`<br>
`JOIN WaterSource ON WaterQuality.SourceID = WaterSource.SourceID`<br>
`GROUP BY CityName;`<br>

In [78]:
cursor.execute("""SELECT CityName, AVG(pH) AS AvgpH,
    RANK() OVER (ORDER BY AVG(pH) DESC) AS pH_Rank
FROM WaterQuality
JOIN WaterSource ON WaterQuality.SourceID = WaterSource.SourceID
JOIN City ON WaterSource.CityID = City.CityID
GROUP BY CityName""")
cursor.fetchall()

[('Kinshasa', 7.568750000000001, 1),
 ('Conakry', 7.559999999999999, 2),
 ('Dar es salaam', 7.514583333333332, 3),
 ("N'djamena", 7.511666666666667, 4),
 ('Antananarivo', 7.497083333333333, 5),
 ('Bamako', 7.406666666666665, 6),
 ('Ouagadougou', 7.4054166666666665, 7),
 ('Ibadan', 7.4004166666666675, 8),
 ('Mbuji-mayi', 7.387916666666666, 9),
 ('Monrovia', 7.376666666666666, 10),
 ('Lilongwe', 7.350833333333334, 11),
 ('Maseru', 7.350833333333333, 12),
 ('Yaoundé', 7.335833333333334, 13),
 ('Kisangani', 7.323750000000001, 14),
 ('Johannesburg', 7.322083333333334, 15),
 ('Kano', 7.319583333333331, 16),
 ('Port harcourt', 7.316666666666669, 17),
 ('Luanda', 7.316250000000004, 18),
 ('Maputo', 7.314166666666669, 19),
 ('Gaborone', 7.305000000000001, 20),
 ('Abuja', 7.297916666666667, 21),
 ('Soweto', 7.2825, 22),
 ('Cairo', 7.258333333333333, 23),
 ('Mombasa', 7.250833333333333, 24),
 ('Lubumbashi', 7.246666666666666, 25),
 ('Accra', 7.242916666666667, 26),
 ('Lagos', 7.240833333333335, 2

### Question 14
A water distribution company wants to identify the water sources with the highest and lowest volumes of water usage per month compared to the average volume across all sources. They need to calculate the deviation of each water source's volume from the average volume and order the results by the testing date and the water source.

Which SQL query correctly calculates the deviation of each water source's volume from the average volume across all sources?

### Options

* `SELECT TestDate, SourceName, Volume - AVG(Volume) OVER () AS VolumeDeviation`<br>
    `FROM WaterUsage`<br>
    `JOIN WaterSource ON WaterUsage.SourceID = WaterSource.SourceID`<br>
    `JOIN City ON WaterSource.CityID = City.CityID`<br>
    `ORDER BY TestDate, SourceName;`

* `SELECT TestDate, SourceName, Volume - AVG(Volume) OVER () AS VolumeDeviation`<br>
`FROM WaterUsage`<br>
`JOIN WaterSource ON WaterUsage.SourceID = WaterSource.SourceID`<br>
`JOIN City ON WaterSource.CityID = City.CityID;`<br>

* `SELECT TestDate, SourceName, 1/Volume * 100 OVER () AS VolumeDeviation`<br>
`FROM WaterUsage`<br>
`JOIN WaterSource ON WaterUsage.SourceID = WaterSource.SourceID`<br>
`JOIN City ON WaterSource.CityID = City.CityID`<br>
`ORDER BY TestDate, SourceName;`

* `SELECT SourceName, Volume - AVG(Volume) OVER (PARTITION BY WaterSource.SourceID) AS VolumeDeviation`<br>
`FROM WaterUsage`<br>
`JOIN WaterSource ON WaterUsage.SourceID = WaterSource.SourceID`<br>
`ORDER BY TestDate, SourceName;`


In [83]:
cursor.execute("""SELECT TestDate, SourceName, Volume - AVG(Volume) OVER () AS VolumeDeviation
FROM WaterUsage
JOIN WaterSource ON WaterUsage.SourceID = WaterSource.SourceID
JOIN City ON WaterSource.CityID = City.CityID
ORDER BY TestDate, SourceName""")
cursor.fetchall()

[('2023-01-01 00:00:00', 'Abidjan431', 5744.406354166727),
 ('2023-01-01 00:00:00', 'Abidjan431', 30270.22635416672),
 ('2023-01-01 00:00:00', 'Abuja897', -1198.8936458332755),
 ('2023-01-01 00:00:00', 'Abuja897', -33078.04364583328),
 ('2023-01-01 00:00:00', 'Accra301', 16556.12635416673),
 ('2023-01-01 00:00:00', 'Accra301', 37884.436354166726),
 ('2023-01-01 00:00:00', 'Addis Ababa596', -5436.743645833274),
 ('2023-01-01 00:00:00', 'Addis Ababa596', 16450.996354166724),
 ('2023-01-01 00:00:00', 'Alexandria703', -12286.233645833272),
 ('2023-01-01 00:00:00', 'Alexandria703', 43134.23635416673),
 ('2023-01-01 00:00:00', 'Antananarivo235', 40157.73635416673),
 ('2023-01-01 00:00:00', 'Antananarivo235', 30922.636354166723),
 ('2023-01-01 00:00:00', 'Bamako767', 19372.996354166724),
 ('2023-01-01 00:00:00', 'Bamako767', -9718.863645833277),
 ('2023-01-01 00:00:00', 'Beira887', -19162.963645833275),
 ('2023-01-01 00:00:00', 'Beira887', 26659.58635416672),
 ('2023-01-01 00:00:00', 'Bloemfo

### Question 15

A water distribution company wants to identify trends in water usage across different regions. They need to calculate the average water usage for each region and determine whether the usage is increasing, decreasing, or remaining stable over the period `1 October 2023 to 31 December 2023`.

Which statement is true about the water usage for the Southeast region over the period?

#### Options
* Water usage in the Southeast region has been increasing.
* Water usage in the Southeast region has been decreasing.
* Water usage in the Southeast region has been stable.
* Water usage in the Southeast region has been varied.

In [127]:
cursor.execute("""SELECT DISTINCT TestDate, Regionname, AVG(Volume) OVER (PARTITION BY Testdate) AS [Average Volume]
FROM WaterUsage
JOIN WaterSource ON WaterUsage.SourceID = WaterSource.SourceID
JOIN City ON WaterSource.CityID = City.CityID
JOIN Region ON City.regionid = region.regionid
WHERE Regionname = 'Southeast' AND testdate BETWEEN '2023-10-01' AND '2023-12-31'
ORDER BY TestDate, regionname""")
cursor.fetchall()

[('2023-10-01 00:00:00', 'Southeast', 62297.4525),
 ('2023-11-01 00:00:00', 'Southeast', 59991.20375000001),
 ('2023-12-01 00:00:00', 'Southeast', 52607.51625)]

### Question 16
A water management agency wants to standardise the format of city names in their database. They need to update all city names to have the first letter capitalised and the remaining letters in lowercase.

Which SQL query correctly updates the city names to have the first letter capitalised and the remaining letters in lowercase?

#### Options
* `UPDATE City`<br>
`SET CityName = UPPER(LEFT(CityName, 1)) || LOWER(RIGHT(CityName, LENGTH(CityName) - 1));` <br>
<br>

* `UPDATE City`<br>
`SET CityName = INITCAP(CityName);`<br>
<br>

* `UPDATE City`<br>
`SET CityName = UPPER(LEFT(CityName, 1)) || LOWER(SUBSTRING(CityName, 2));`<br>
<br>

* `UPDATE City`<br>
`SET CityName = UPPER(SUBSTR(CityName, 1, 1)) || LOWER(SUBSTR(CityName, 2));`

In [100]:
cursor.execute("""SELECT Cityname, UPPER(SUBSTR(CityName, 1, 1)) || LOWER(SUBSTR(CityName, 2)) FROM City""")
cursor.fetchall()

[('Cairo', 'Cairo'),
 ('Lagos', 'Lagos'),
 ('Kinshasa', 'Kinshasa'),
 ('Giza', 'Giza'),
 ('Nairobi', 'Nairobi'),
 ('Johannesburg', 'Johannesburg'),
 ('Abidjan', 'Abidjan'),
 ('Casablanca', 'Casablanca'),
 ('Cape town', 'Cape town'),
 ('Durban', 'Durban'),
 ('Alexandria', 'Alexandria'),
 ('Addis ababa', 'Addis ababa'),
 ('Dar es salaam', 'Dar es salaam'),
 ('Accra', 'Accra'),
 ('Khartoum', 'Khartoum'),
 ('Suez', 'Suez'),
 ('Mombasa', 'Mombasa'),
 ('Kampala', 'Kampala'),
 ('Maputo', 'Maputo'),
 ('Port harcourt', 'Port harcourt'),
 ('Ouagadougou', 'Ouagadougou'),
 ('Lusaka', 'Lusaka'),
 ('Abuja', 'Abuja'),
 ('Bamako', 'Bamako'),
 ('Kano', 'Kano'),
 ('Dakar', 'Dakar'),
 ('Cotonou', 'Cotonou'),
 ('Lilongwe', 'Lilongwe'),
 ('Ibadan', 'Ibadan'),
 ('Luanda', 'Luanda'),
 ('Yaoundé', 'Yaoundé'),
 ('Antananarivo', 'Antananarivo'),
 ('Maiduguri', 'Maiduguri'),
 ('Harare', 'Harare'),
 ('Conakry', 'Conakry'),
 ('Soweto', 'Soweto'),
 ('Kisangani', 'Kisangani'),
 ('Mbuji-mayi', 'Mbuji-mayi'),
 ('Maser

### Question 17

Furthermore, the water management agency wants to create an area tag that uniquely identifies each area based on the country, region, and city. 

The area tag should consist of the first letter of the first word and the first letter of the second word if the **country name consists of two words**, both capitalised. **Otherwise**, it should include the first two letters of the country name, with the first letter capitalised and the second letter in lowercase. 

**Additionally, for the region and city parts**, the first letter should be capitalised and the subsequent letters should be in lowercase. 

Each part of the area tag should be **separated by a dash**.

**Example:**
An area in South Africa, in the South region and in the city of Johannesburg, would be tagged **SA-Sou-Joh**, while an area in Nigeria, in the eastern region and in the city of Lagos, would be tagged **Ni-Eas-Lag**. 

After constructing the area tags according to the provided instructions, order the results in descending order according to the area tag.

Which three tags appear first in the list?


### Options
* Be-Eas-Cai, Be-Eas-Lag, CV-Nor-Kin
* Al-Nor-Abi, Al-Nor-Acc, Al-Nor-Cot
* Benin-Eas-Cai, Benin-Eas-Lag, Cabo verde-Nor-Kin
* Ca-Sou-Kam, Ca-Sou-Cap, Ca-Sou-Blo


In [113]:
cursor.execute("""SELECT Countryname, Regionname, Cityname, UPPER(SUBSTR(Countryname,1,2))|| '-' ||SUBSTR(Regionname,1,3)||'-'||SUBSTR(Cityname,1,3) AS Area
FROM City 
JOIN Region ON city.regionid = region.regionid
JOIN Country ON city.countryid = country.countryid
ORDER BY Area
""")
cursor.fetchall()

[('Algeria', 'West', 'Alexandria', 'AL-Wes-Ale'),
 ('Angola', 'East', 'Casablanca', 'AN-Eas-Cas'),
 ('Angola', 'East', 'Lagos', 'AN-Eas-Lag'),
 ('Angola', 'South', 'Johannesburg', 'AN-Sou-Joh'),
 ('Benin', 'Northeast', 'Dar es salaam', 'BE-Nor-Dar'),
 ('Benin', 'Northwest', 'Monrovia', 'BE-Nor-Mon'),
 ('Burundi', 'East', 'Accra', 'BU-Eas-Acc'),
 ('Burundi', 'East', 'Kisangani', 'BU-Eas-Kis'),
 ('Burkina Faso', 'Southeast', 'Cape town', 'BU-Sou-Cap'),
 ('Cameroon', 'East', 'Kano', 'CA-Eas-Kan'),
 ('Cameroon', 'West', 'Gaborone', 'CA-Wes-Gab'),
 ('Central African Republic', 'West', 'Durban', 'CE-Wes-Dur'),
 ('Congo', 'East', 'Soweto', 'CO-Eas-Sow'),
 ('Comoros', 'Northwest', 'Beira', 'CO-Nor-Bei'),
 ('Comoros', 'Northeast', 'Mbuji-mayi', 'CO-Nor-Mbu'),
 ('Djibouti', 'Northwest', 'Kumasi', 'DJ-Nor-Kum'),
 ('Djibouti', 'South', 'Freetown', 'DJ-Sou-Fre'),
 ('Egypt', 'Northeast', 'Kinshasa', 'EG-Nor-Kin'),
 ('Egypt', 'Northwest', 'Lilongwe', 'EG-Nor-Lil'),
 ('Equatorial Guinea', 'East', 'Lub

In [21]:
cursor.execute("""SELECT
    IFNULL(CASE
            WHEN LENGTH(Country.CountryName) - LENGTH(REPLACE(Country.CountryName, ' ', '')) = 1 THEN UPPER(SUBSTR(Country.CountryName, 1, 1)) || UPPER(SUBSTR(SUBSTR(Country.CountryName, INSTR(Country.CountryName, ' ') + 1), 1, 1))
            ELSE UPPER(SUBSTR(Country.CountryName, 1, 1)) || LOWER(SUBSTR(Country.CountryName, 2, 1))
        END, '') || '-' ||
    UPPER(SUBSTR(Region.RegionName, 1, 1)) || '-' ||
    UPPER(SUBSTR(City.CityName, 1, 1)) AS AreaTag
FROM
    City
JOIN
    Region ON City.RegionID = Region.RegionID
JOIN
    Country ON Region.CountryID = Country.CountryID
ORDER BY
    AreaTag DESC
""")
cursor.fetchall()

[('Ca-S-K',),
 ('Ca-S-C',),
 ('Ca-S-B',),
 ('Ca-S-A',),
 ('CV-N-M',),
 ('CV-N-M',),
 ('CV-N-M',),
 ('CV-N-L',),
 ('CV-N-K',),
 ('CV-N-D',),
 ('CV-N-D',),
 ('CV-N-A',),
 ('Bu-S-B',),
 ('Bu-S-A',),
 ('Bo-W-P',),
 ('Bo-W-N',),
 ('Bo-W-L',),
 ('Bo-W-K',),
 ('Bo-W-G',),
 ('Bo-W-D',),
 ('Bo-W-C',),
 ('Bo-W-A',),
 ('Bo-W-A',),
 ('Be-E-S',),
 ('Be-E-O',),
 ('Be-E-N',),
 ('Be-E-L',),
 ('Be-E-L',),
 ('Be-E-K',),
 ('Be-E-K',),
 ('Be-E-I',),
 ('Be-E-C',),
 ('Be-E-C',),
 ('Be-E-B',),
 ('Be-E-A',),
 ('BF-N-S',),
 ('BF-N-M',),
 ('BF-N-M',),
 ('BF-N-L',),
 ('BF-N-K',),
 ('BF-N-B',),
 ('An-S-M',),
 ('An-S-J',),
 ('An-S-G',),
 ('An-S-F',),
 ('Al-N-Y',),
 ('Al-N-H',),
 ('Al-N-C',)]

In [ ]:
|| '-' ||LEFT(Regionname,3)||'-'||LEFT(Cityname,3)

### Question 18

The water treatment facility management team wants to create a stored procedure to update the capacity of a water treatment plant in the database. They have started writing the code but have found it incomplete.

Which line of code is missing from the following SQL code snippet to complete the stored procedure "UpdatePlantCapacity"?

#### Options
* `ROLLBACK;`<br>
Missing after the BEGIN HANDLER block to rollback the transaction in case of an error.
<br>
<br>
* `PRINT 'Error updating plant capacity.';`<br>
Missing after the BEGIN HANDLER block to print an error message in case of an error.
<br>
<br>
* `THROW;` <br> 
Missing after the BEGIN HANDLER block to re-throw the error and terminate the procedure.
<br>
<br>
* `SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Error updating plant capacity.';`<br>
Missing after the BEGIN HANDLER block to raise a custom error in case of an error.

In [116]:
"""
CREATE PROCEDURE UpdatePlantCapacity(
    IN p_PlantID INT,
    IN p_NewCapacity FLOAT
)
BEGIN
    DECLARE exit handler for SQLEXCEPTION
    BEGIN
        ROLLBACK;
    END;

    START TRANSACTION;
        UPDATE WaterTreatmentPlant
        SET Capacity = p_NewCapacity
        WHERE PlantID = p_PlantID;
    COMMIT;
END
SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Error updating plant capacity.';

"""

"\nCREATE PROCEDURE UpdatePlantCapacity(\n    IN p_PlantID INT,\n    IN p_NewCapacity FLOAT\n)\nBEGIN\n    DECLARE exit handler for SQLEXCEPTION\n    BEGIN\n        ROLLBACK;\n    END;\n\n    START TRANSACTION;\n        UPDATE WaterTreatmentPlant\n        SET Capacity = p_NewCapacity\n        WHERE PlantID = p_PlantID;\n    COMMIT;\nEND\nSIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'Error updating plant capacity.';\n\n"

### Question 19 
A data entry error has resulted in incorrect pH values being inserted into the WaterQuality table for the city of Ouagadougou. The agency wants to correct these errors and ensure that the corrections are committed only if all updates are successful.

Which of the following SQL code snippets correctly demonstrates the use of a transaction with SAVEPOINT, ROLLBACK, and COMMIT to correct the pH values in the WaterQuality table for the city of **Ouagadougou**?

### Options


* `BEGIN TRANSACTION;`<br>
`UPDATE WaterQuality`<br>
`SET pH = 7.24`<br>
&nbsp;&nbsp;&nbsp;&nbsp;`WHERE SourceID IN (SELECT SourceID FROM WaterSource WHERE CityID = (SELECT CityID FROM City WHERE CityName = 'Ouagadougou'));`<br>
`ROLLBACK;`<br>
<br>
<br>
* `BEGIN TRANSACTION;`<br>
`SAVEPOINT data_correction;`<br>
`UPDATE WaterQuality`<br>
`SET pH = 7.24`<br>
&nbsp;&nbsp;&nbsp;&nbsp;`WHERE SourceID IN (SELECT SourceID FROM WaterSource WHERE CityID = (SELECT CityID FROM City WHERE CityName = 'Ouagadougou'));`<br>
`COMMIT;`<br>
<br>
<br>
* `BEGIN TRANSACTION;`<br>
`SAVEPOINT data_correction;`<br>
`UPDATE WaterQuality`<br>
`SET pH = 7.24`<br>
&nbsp;&nbsp;&nbsp;&nbsp;`WHERE SourceID IN (SELECT SourceID FROM WaterSource WHERE CityID = (SELECT CityID FROM City WHERE CityName = 'Ouagadougou'));`<br>
`ROLLBACK TO SAVEPOINT data_correction;`<br>
`COMMIT;` <br>
<br>
<br>
* `BEGIN TRANSACTION;`<br>
`UPDATE WaterQuality`<br>
`SET pH = 7.24`<br>
&nbsp;&nbsp;&nbsp;&nbsp;`WHERE SourceID IN (SELECT SourceID FROM WaterSource WHERE CityID = (SELECT CityID FROM City WHERE CityName = 'Ouagadougou'));`<br>
`ROLLBACK TO SAVEPOINT data_correction;`<br>
`COMMIT;`<br> 
`IN (SELECT SourceID FROM WaterSource WHERE CityID = (SELECT CityID FROM City WHERE CityName = 'Ouagadougou'));`<br>
`ROLLBACK TO SAVEPOINT data_correction;`<br>
`COMMIT;`<br>

In [122]:
cursor.execute("""

UPDATE WaterQuality
SET pH = 7.24
    WHERE SourceID IN (SELECT SourceID FROM WaterSource WHERE CityID = (SELECT CityID FROM City WHERE CityName = 'Ouagadougou'))""")
cursor.fetchall()

[]

In [121]:
cursor.execute("""BEGIN TRANSACTION;
UPDATE WaterQuality
SET pH = 7.24
    WHERE SourceID IN (SELECT SourceID FROM WaterSource WHERE CityID = (SELECT CityID FROM City WHERE CityName = 'Ouagadougou'));
ROLLBACK TO SAVEPOINT data_correction;
COMMIT;
IN (SELECT SourceID FROM WaterSource WHERE CityID = (SELECT CityID FROM City WHERE CityName = 'Ouagadougou'));
ROLLBACK TO SAVEPOINT data_correction;
COMMIT;
""")
cursor.fetchall()

ProgrammingError: You can only execute one statement at a time.

### Question 20

The water management agency wants to analyse the total population for each country-region combination. They need to concatenate the country name with the region name and calculate the total population for each country-region after grouping by country-region.

If we order the result by the total population in descending order, which country-region combination has the highest population?

### Options
* Burundi
* Benin - East
* Burundi - Southwest	
* Benin

In [14]:
cursor.execute("""SELECT 
    Country.CountryName || '-' || Region.RegionName AS CountryRegion,
    SUM(City.Population) AS TotalPopulation
FROM 
    City
JOIN 
    Region ON City.RegionID = Region.RegionID
JOIN 
    Country ON Region.CountryID = Country.CountryID
GROUP BY 
    CountryRegion
ORDER BY 
    TotalPopulation DESC
LIMIT 1
""")
cursor.fetchall()

[('Benin-East', 80739559)]

In [ ]:
cursor.execute("""SELECT (countryname,regionname), SUM(Population) AS Total
FROM Country 
JOIN Region ON COUNTRY.COUNTRYid = region.COUNTRYid
JOIN City ON city.regionid = region.regionid
ORDER BY Total DESC
""")
cursor.fetchall()

In [6]:
cursor.execute("""SELECT 
    Country.CountryName || '-' || Region.RegionName AS CountryRegion,
    SUM(City.Population) AS TotalPopulation
FROM 
    City
JOIN 
    Region ON City.RegionID = Region.RegionID
JOIN 
    Country ON Region.CountryID = Country.CountryID
GROUP BY 
    CountryRegion
ORDER BY 
    TotalPopulation DESC 
    LIMIT 1
""")
cursor.fetchall()

[('Benin-East', 80739559)]

### Question 21
The water quality testing agency wants to analyse the water quality data for both 2023 and 2024. They need to combine the data from the "WaterQuality" table (for 2023) with the data from the "WaterQuality24" table (for 2024) using a union operation.

After performing the union, what is the average pH and ChlorineLevel over 2023 and 2024.

### Options
* 6.34, 0,87
* 7.5, 0.76
* 7.75, 0.97
* None of the above

In [15]:
#Your code here
cursor.execute("""SELECT AVG(pH) AS AvgpH, AVG(ChlorineLevel) AS AvgChlorineLevel
FROM (
    SELECT pH, ChlorineLevel FROM WaterQuality
    UNION ALL
    SELECT pH, ChlorineLevel FROM WaterQuality24
) AS CombinedData
""")
cursor.fetchall()

[(7.7556293402777925, 0.9660850694444432)]

### Question 22
The water quality control department wants to analyse water quality data for cities with a population greater than 1 million and a pH level higher than 7. They need to filter the data to include only the relevant records based on these conditions.

Based on these filters, which quality test (QualityID) has the highest pH and ChlorineLevels?

### Options
* 1
* 688
* 572
* 241

In [16]:
#Your code here
cursor.execute("""SELECT 
    QualityID,
    MAX(pH) AS MaxpH,
    MAX(ChlorineLevel) AS MaxChlorineLevel
FROM 
    WaterQuality
JOIN 
    WaterSource ON WaterQuality.SourceID = WaterSource.SourceID
JOIN 
    City ON WaterSource.CityID = City.CityID
WHERE 
    City.Population > 1000000
    AND WaterQuality.pH > 7
GROUP BY 
    QualityID
ORDER BY 2,3 DESC
""")
cursor.fetchall()


[(688, 7.01, 1.47),
 (213, 7.01, 0.76),
 (572, 7.01, 0.27),
 (899, 7.02, 1.08),
 (295, 7.02, 1.05),
 (68, 7.03, 0.41),
 (1004, 7.03, 0.37),
 (874, 7.04, 1.34),
 (851, 7.04, 0.89),
 (675, 7.04, 0.26),
 (171, 7.05, 0.8),
 (509, 7.05, 0.61),
 (127, 7.06, 1.17),
 (764, 7.06, 0.94),
 (367, 7.06, 0.93),
 (61, 7.06, 0.75),
 (1115, 7.06, 0.41),
 (54, 7.06, 0.29),
 (637, 7.07, 1.49),
 (181, 7.07, 1.46),
 (739, 7.07, 1.41),
 (126, 7.07, 1.29),
 (294, 7.07, 0.8),
 (330, 7.07, 0.39),
 (1033, 7.08, 1.26),
 (366, 7.08, 0.77),
 (945, 7.08, 0.5),
 (690, 7.08, 0.24),
 (856, 7.09, 1.37),
 (191, 7.09, 1.19),
 (1100, 7.09, 1.09),
 (64, 7.09, 0.77),
 (168, 7.09, 0.51),
 (275, 7.1, 1.42),
 (448, 7.1, 0.63),
 (767, 7.1, 0.35),
 (524, 7.11, 0.82),
 (1113, 7.11, 0.38),
 (60, 7.11, 0.29),
 (128, 7.12, 1.47),
 (745, 7.12, 1.45),
 (716, 7.12, 1.29),
 (132, 7.12, 0.85),
 (271, 7.13, 1.35),
 (478, 7.13, 1.34),
 (500, 7.13, 1.27),
 (266, 7.13, 0.69),
 (402, 7.13, 0.48),
 (53, 7.13, 0.4),
 (731, 7.14, 1.47),
 (869, 7

### Question 23

The water quality control department wants to classify water sources based on their pH levels. They need to categorise water sources as "Low pH" if the pH level is less than 7, "Medium pH" if the pH level is between 7 and 7.5 inclusive, and "High pH" if the pH level is greater than 7.5.

What is the pH category for the **Cairo839** source?

### Options
* Medium pH
* Low pH
* High pH
* None of the above


In [17]:
#Your code here
cursor.execute("""SELECT
    ws.sourcename,
    wq.ph,
    CASE
        WHEN wq.ph < 7 THEN 'Low pH'
        WHEN wq.ph BETWEEN 7 AND 7.5 THEN 'Medium pH'
        WHEN wq.ph > 7.5 THEN 'High pH'
    END AS pH_category
FROM
    WaterSource ws
JOIN
    WaterQuality wq ON ws.sourceid = wq.sourceid
WHERE
    ws.sourcename = 'Cairo839'
""")
cursor.fetchall()

[('Cairo839', 7.39, 'Medium pH'),
 ('Cairo839', 7.66, 'High pH'),
 ('Cairo839', 6.68, 'Low pH'),
 ('Cairo839', 8.18, 'High pH'),
 ('Cairo839', 7.85, 'High pH'),
 ('Cairo839', 6.08, 'Low pH'),
 ('Cairo839', 6.77, 'Low pH'),
 ('Cairo839', 7.91, 'High pH'),
 ('Cairo839', 6.44, 'Low pH'),
 ('Cairo839', 6.73, 'Low pH'),
 ('Cairo839', 8.19, 'High pH'),
 ('Cairo839', 7.8, 'High pH'),
 ('Cairo839', 7.22, 'Medium pH'),
 ('Cairo839', 7.65, 'High pH'),
 ('Cairo839', 8.21, 'High pH'),
 ('Cairo839', 6.58, 'Low pH'),
 ('Cairo839', 7.54, 'High pH'),
 ('Cairo839', 6.63, 'Low pH'),
 ('Cairo839', 7.62, 'High pH'),
 ('Cairo839', 6.24, 'Low pH'),
 ('Cairo839', 7.57, 'High pH'),
 ('Cairo839', 6.07, 'Low pH'),
 ('Cairo839', 6.79, 'Low pH'),
 ('Cairo839', 8.4, 'High pH')]

### Question 24
The water treatment facility management team wants to implement a stored procedure to update the capacity of water treatment plants. They need to handle potential errors gracefully using error handling techniques in MySQL stored procedures and provide informative error messages in case of failures.

Which SQL code snippet correctly creates a stored procedure named "UpdatePlantCapacity" to update the capacity of a water treatment plant and handles potential errors?

### Options

* `CREATE PROCEDURE UpdatePlantCapacity(`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`IN PlantID INT,`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`IN NewCapacity REAL`<br>
&nbsp;&nbsp;&nbsp;&nbsp;`)`<br>
`BEGIN`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`UPDATE WaterTreatmentPlant`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`SET Capacity = NewCapacity`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`WHERE PlantID = PlantID;`<br><br>
    &nbsp;&nbsp;&nbsp;&nbsp;`SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'An attempt to signal an error without a handler.';`<br>
`END;`<br>
<br>
<br>
* `CREATE PROCEDURE UpdatePlantCapacity(`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`IN PlantID INT,`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`IN NewCapacity REAL`<br>
`)`<br>
`BEGIN`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`UPDATE WaterTreatmentPlant`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`SET Capacity = NewCapacity`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`WHERE PlantID = PlantID;`<br><br>
    &nbsp;&nbsp;&nbsp;&nbsp;`SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'An attempt to signal an error without a handler.';`<br>
`END;`<br>
<br>
<br>
* `CREATE PROCEDURE UpdatePlantCapacity(`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`IN PlantID INT,`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`IN NewCapacity REAL`<br>
&nbsp;&nbsp;&nbsp;&nbsp;`)`<br>
`BEGIN`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`UPDATE WaterTreatmentPlant`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`SET Capacity = NewCapacity`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`WHERE PlantID = PlantID;`<br>
`END;`<br>
<br>
<br>
* `CREATE PROCEDURE UpdatePlantCapacity(`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`IN PlantID INT,`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`IN NewCapacity REAL`<br>
&nbsp;&nbsp;&nbsp;`)`<br>
`BEGIN`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`DECLARE EXIT HANDLER FOR SQLEXCEPTION`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`BEGIN`<br>
        &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;`SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'An error occurred while updating plant capacity.';`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`END;`<br><br>
    &nbsp;&nbsp;&nbsp;&nbsp;`UPDATE WaterTreatmentPlant`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`SET Capacity = NewCapacity`<br>
    &nbsp;&nbsp;&nbsp;&nbsp;`WHERE PlantID = PlantID;`<br>
`END;`


### Question 25
The water treatment facility management team wants to create a stored procedure to insert a new water treatment plant into the database. They need to ensure that if the insertion fails due to a **primary key violation**, the procedure gracefully handles the error and provides a custom error message.

A data analyst is struggling to create the procedure to fulfil this task, what is wrong with his code?


#### Options

* The code block does not handle errors properly because it does not use the correct error code for primary key violations.
* The code block does not handle errors properly because it checks for a plant with the same name already existing, rather than a primary key violation based on the PlantID.
* The code block does not handle errors properly because it does not include a condition handler for the primary key violation based on the PlantID.
* The code block does not handle errors properly because it does not provide a detailed error message.

In [ ]:
"""
DELIMITER //

CREATE PROCEDURE InsertNewPlant
    (IN p_PlantID INT,
     IN p_PlantName NVARCHAR(255),
     IN p_Capacity FLOAT)
BEGIN
    DECLARE duplicate_key CONDITION FOR SQLSTATE '23000';
    DECLARE CONTINUE HANDLER FOR duplicate_key
        BEGIN
            SIGNAL SQLSTATE '45000' SET MESSAGE_TEXT = 'A plant with the same ID already exists.';
        END;
        
    INSERT INTO WaterTreatmentPlant (PlantID, PlantName, Capacity)
    VALUES (p_PlantID, p_PlantName, p_Capacity);
    
END //

DELIMITER ;

"""

### Question 26
A water quality testing agency wants to update the pH values of water samples in the WaterQuality table for a specific city and a certain month. However, they want to ensure that all updates are either successfully committed or rolled back if any errors occur during the process.

Write a query that uses a transaction with COMMIT and ROLLBACK to update the pH value in the WaterQuality table for the city of "Ouagadougou" for January 2023 to -12.

What is the updated average pH for the city of "Ouagadougou" now?

**Hint: In order to execute transactions in this notebook, we need to use sqlite3 commands.**

**This query modifies a value in the dataset. Consequently, to rerun the query or queries from previous questions, please download and use a fresh copy of the dataset.**

### Options
* 5.74
* 7.74
* 7.41
* -12

In [18]:
#Your code here
import sqlite3
 
# Establish connection and create a cursor
conn = sqlite3.connect('/mnt/data/water_quality.db')  # Replace with the actual path to your database file
cur = conn.cursor()
 
# Step 1: Retrieve source IDs
cur.execute("""
    SELECT ws.sourceid
    FROM WaterSource ws
    JOIN City c ON ws.cityid = c.cityid
    WHERE c.cityname = 'Ouagadougou'
""")
source_ids = cur.fetchall()
source_ids = tuple([id[0] for id in source_ids])
 
# Step 2: Update WaterQuality records
try:
    conn.execute("BEGIN TRANSACTION;")
    update_query = f"""
        UPDATE WaterQuality
        SET ph = -12
        WHERE sourceid IN {source_ids}
        AND strftime('%Y-%m', testdate) = '2023-01';
    """
    conn.execute(update_query)
    # Assuming successful execution
    conn.execute("COMMIT;")
except Exception as e:
    print("Error occurred:", e)
    conn.execute("ROLLBACK;")
 
# Step 3: Calculate the updated average pH
cur.execute("""
    SELECT AVG(wq.ph) AS average_ph
    FROM WaterQuality wq
    JOIN WaterSource ws ON wq.sourceid = ws.sourceid
    JOIN City c ON ws.cityid = c.cityid
    WHERE c.cityname = 'Ouagadougou'
""")
average_ph = cur.fetchone()[0]
 
# Close the connection
conn.close()

OperationalError: unable to open database file

#  

<div align="center" style=" font-size: 80%; text-align: center; margin: 0 auto">
<img src="https://raw.githubusercontent.com/Explore-AI/Pictures/master/ExploreAI_logos/EAI_Blue_Dark.png"  style="width:200px";/>
</div>